### PR TITLE
Added a login route

### PR DESCRIPTION
**Pull request by AdamAkiva**
Added the following:
* Login route and everything relating to it (api docs & tests as well)
* Added [argon2](https://github.com/ranisalt/node-argon2) password hashing when creating/updating a user.
* When OAuth2.0 will be added, this route will be extended to allow that as well

### CODE REVIEW

File: scripts/docker-compose.yml [hunk 1]
```diff
@@ -108,14 +108,13 @@ services:
       - HTTP_ROUTE=v0/api
       - WEB_SOCKET_ROUTE=v0/sockets
       - HEALTH_CHECK_ROUTE=health
-      - ALLOWED_HOSTS=localhost
+      - ALLOWED_HOSTS=::,127.0.0.1,localhost
```
Comment (Meital-Kozhidov) [inline scripts/docker-compose.yml:111]: Why?

File: be/src/utils/logger.ts [hunk 2]
```diff
@@ -84,11 +83,11 @@ export default class Logger {
     this._handler = this._logMiddleware.logger;
   }
 
-  public get logMiddleware() {
```
Comment (Meital-Kozhidov) [inline be/src/utils/logger.ts:87]: Why the change?

File: be/src/server/http.ts [hunk 3]
```diff
@@ -120,11 +144,17 @@ export default class HttpServer {
 
   public async attachConfigurationMiddlewares() {
     this._app.use(
-      Middlewares.checkMethod(HttpServer.ALLOWED_METHODS),
+      Middlewares.checkMethod(this._allowedMethods),
       cors(this._corsOptions),
       compress()
     );
     if (isProductionMode(this._mode)) {
+      // TODO When going to production, match the value to the number of reverse
```
Comment (Meital-Kozhidov) [inline be/src/server/http.ts:152]: :O

File: be/src/utils/logger.ts (hunk 2 above)
Comment (AdamAkiva) [reply be/src/utils/logger.ts:87]: Easier to mock the function when it is a function.
`getter` or `setter` are considered as variables by the test library, and function mocks are easier to implement

File: be/db/migrations/0000_wealthy_miss_america.sql [hunk 4]
```diff
@@ -2,7 +2,7 @@ CREATE TABLE IF NOT EXISTS "users_credentials" (
 	"id" uuid PRIMARY KEY DEFAULT gen_random_uuid() NOT NULL,
 	"user_id" uuid NOT NULL,
 	"email" varchar(256) NOT NULL,
-	"password" varchar(64) NOT NULL,
+	"password" varchar(255) NOT NULL,
```
Comment (Meital-Kozhidov) [inline be/db/migrations/0000_wealthy_miss_america.sql:5]: Why is the change and why 255 and not 256?

File: be/src/db/handler.ts [hunk 5]
```diff
@@ -126,6 +128,10 @@ export default class Database {
     return await this._conn.end({ timeout: 10 }); // in secs
   }
 
+  public async isReady() {
```
Comment (Meital-Kozhidov) [inline be/src/db/handler.ts:131]: I liked the isReady encapsulation in the different modules

File: be/src/db/schemas.ts [hunk 6]
```diff
@@ -104,11 +103,11 @@ export const userCredentialsModel = pgTable(
...
       ),
-    email: varchar('email', { length: USER_EMAIL_MAX_LENGTH })
-      .unique()
-      .notNull(),
+    email: varchar('email', { length: EMAIL_MAX_LENGTH }).unique().notNull(),
     password: varchar('password', {
-      length: USER_PASSWORD_MAX_LENGTH
+      // The password is hashed so the allowed length is larger than what we allow
+      // the user to set
+      length: 255
```
Comment (Meital-Kozhidov) [inline be/src/db/schemas.ts:110]: So I got an answer about the enlargement here

File: be/src/db/schemas.ts [hunk 7]
```diff
@@ -118,9 +117,11 @@ export const userCredentialsModel = pgTable(
     // may send a query params to return ONLY archived users instead.
     // A read of soft-deleted user by id will work.
 
-    // A creation of a soft-deleted user will fail with an existing email message.
+    // A creation of a soft-deleted user will fail with a 409 (conflict)
```
Comment (Meital-Kozhidov) [inline be/src/db/schemas.ts:120]: This is interesting - why doesn't it activate it instead?

File: be/src/entities/auth/route.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+} from '../../utils/index.js';
+
+import { authEmailPass } from './controller.js';
+
+/**********************************************************************************/
+
+export function authRouter(rateLimiter: RateLimitRequestHandler) {
+  return Router().post(
+    '/auth',
+    json({ limit: '8kb' }),
```
Comment (Meital-Kozhidov) [inline be/src/entities/auth/route.ts:14]: Different limits are spreaded thoughout the code. Have you considered have all these values in one place?

File: be/src/entities/auth/service/create.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+import { passwordMismatchError, userNotFoundError } from './errors.js';
+
+/**********************************************************************************/
+
+type AuthEmailPassValidationData = ReturnType<typeof authEmailPassValidation>;
+
+/**********************************************************************************/
+
+export async function authEmailPass(
```
Comment (Meital-Kozhidov) [inline be/src/entities/auth/service/create.ts:20]: Not loving this  function name

File: be/src/entities/utils.ts [hunk 10]
```diff
@@ -115,3 +116,11 @@ export async function executePreparedQuery<
...
   >;
 }
+
+export async function hashPassword(password: string) {
+  return await argon2.hash(password);
+}
+
+export async function verifyPassword(hash: string, password: string) {
+  return await argon2.verify(hash, password);
+}
```
Comment (Meital-Kozhidov) [inline be/src/entities/utils.ts:126]: Does this emoji mean 'missing new line in eof'? because seems so

File: be/src/event-emitters/message-queue.ts [hunk 11]
```diff
@@ -118,28 +127,39 @@ export default class MessageQueue {
...
-    });
+      this._channel.publish(exchangeName, routingKey, Buffer.from(data), {
+        expiration: 600_000, // millis
+        persistent: true
+      });
+    } catch (err) {
+      if (err instanceof TypeError) {
+        this._logger.error(
+          err,
+          'An attempt to stringify a BigInt/circular value'
```
Comment (Meital-Kozhidov) [inline be/src/event-emitters/message-queue.ts:150]: Interesting, are you going to trycatch every stringity?

File: be/src/event-emitters/socket-server.ts [hunk 12]
```diff
@@ -92,11 +91,14 @@ export default class WebSocketServer {
 
   private _logClientConnection(socket: Socket) {
     const header = socket.handshake.headers['x-forwarded-for'];
-    let clientIp = socket.handshake.address;
+
+    let clientIp = '';
```
Comment (Meital-Kozhidov) [inline be/src/event-emitters/socket-server.ts:95]: Why did you rewrite this?

File: be/src/init.ts [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    logger: logger,
+    debug: cacheDebug
+  });
+  await cache.connect();
+
+  return cache;
+}
+
+export function createRateLimiter(cache: Cache, logger: Logger['_handler']) {
+  return new RateLimiter(
```
Comment (Meital-Kozhidov) [inline be/src/init.ts:61]: I see there are some configuration fields, did you check them out?

File: be/src/server/http.ts [hunk 14]
```diff
@@ -91,18 +104,29 @@ export default class HttpServer {
...
     if (port) {
       port = typeof port === 'string' ? Number(port) : port;
     }
 
     return await new Promise<number>((resolve) => {
       this._server.once('listening', () => {
-        callback?.();
-        resolve((this._server.address()! as AddressInfo).port);
+        if (!isTestMode(this._mode)) {
+          // Can be asserted since this is not a unix socket and we are inside
```
Comment (Meital-Kozhidov) [inline be/src/server/http.ts:118]: Can you elaborate?

File: be/src/server/http.ts [hunk 15]
```diff
@@ -91,18 +104,29 @@ export default class HttpServer {
...
+          // the listen event
+          const { address, port } = this._server.address() as AddressInfo;
+          const route = this._routes.http;
+
+          this._logger.info(
+            `Server is running in '${this._mode}' mode on: ` +
+              `'${address.endsWith(':') ? address : address.concat(':')}${port}${route}'`
```
Comment (Meital-Kozhidov) [inline be/src/server/http.ts:125]: What's that?

File: be/package.json [hunk 16]
```diff
@@ -30,23 +30,27 @@
...
+    "amqplib": "~0.10.4",
+    "argon2": "~0.40.1",
     "compression": "~1.7.4",
     "cors": "~2.8.5",
     "debug": "~4.3.4",
-    "drizzle-orm": "~0.30.7",
+    "drizzle-orm": "~0.30.8",
     "express": "~4.19.2",
+    "express-rate-limit": "~7.2.0",
+    "fast-json-stable-stringify": "~2.1.0",
```
Comment (Meital-Kozhidov) [inline be/package.json:41]: Why?

File: be/src/utils/rate-limiter.ts [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+  // See this:
+  // https://github.com/express-rate-limit/express-rate-limit/blob/main/docs/guides/troubleshooting-proxy-issues.mdx#port-numbers-in-ip-addresses
+  private _keyGenerator(req: Request) {
+    if (!req.ip) {
+      this._logger.error('IP is missing from the request');
+
+      return req.socket.remoteAddress!;
+    }
+
+    return req.ip;
```
Comment (Meital-Kozhidov) [inline be/src/utils/rate-limiter.ts:63]: In the link they do `return request.ip.replace(/:\d+[^:]*$/, '')` though

File: be/__tests__/integration/auth.test.ts [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+
+    const { data, statusCode } = await sendHttpRequest<string>(authURL, {
+      method: 'post',
+      json: {
+        email: email,
+        password: 'Bla123!@#'
+      } satisfies AuthEmailPass
+    });
+    expect(statusCode).toBe(StatusCodes.SUCCESS);
+    expect(typeof data === 'string').toBe(true);
```
Comment (Meital-Kozhidov) [inline be/__tests__/integration/auth.test.ts:62]: Can this line can written as `expect(typeof data).toBe('string')` ?

File: be/__tests__/integration/general.test.ts [hunk 19]
```diff
@@ -14,11 +17,32 @@ import {
 
 /**********************************************************************************/
 
-describe.runIf(isStandardTest()).concurrent('General tests', () => {
-  beforeAll(initServer);
-  afterAll(terminateServer);
+vi.mock('../../src/entities/utils.js', async (originalImport) => {
```
Comment (Meital-Kozhidov) [inline be/__tests__/integration/general.test.ts:20]: That's actually really cool

File: be/src/utils/rate-limiter.ts (hunk 17 above)
Comment (AdamAkiva) [reply be/src/utils/rate-limiter.ts:63]: Until we go to production with reverse proxy in place it is less relevant.
Since the purpose of this is to remove the port number from the ip address, I don't see a reason to do it.
As they say: "While some reverse proxy pass a comma delimited list of IP address, some proxies (for example, Azure's Application Gateway) will pass a comma delimited list of IP:PORT instead, where the port is the source port, that can change on every request."
I didn't not encounter a use case like this (ever), if we do at some point I'll fix it

File: be/__tests__/integration/auth.test.ts (hunk 18 above)
Comment (AdamAkiva) [reply be/__tests__/integration/auth.test.ts:62]: Probably, any particular reason for asking?

File: be/__tests__/integration/general.test.ts (hunk 19 above)
Comment (AdamAkiva) [reply be/__tests__/integration/general.test.ts:20]: I can try to do the same trick as the event emitter (to import it first) in order to make sure it happens before everything else (`vitest` docs says it is hoisted to the top but I don't know how it will behave when it is imported by other files)
I'm a bit worried about the import path and I felt like having this at the start is more verbose to know exactly what is mocked.
Perhaps I'll try it later, but I'm not sure if I want to actually go in that direction

File: be/__tests__/utils.ts [hunk 20]
```diff
@@ -75,7 +93,42 @@ export type MQUserRoutingKeys = ExchangesAndRoutes['user.exchange'][number];
 /***************************** Server setup ***************************************/
 /**********************************************************************************/
 
-export async function createServer() {
+export async function initServer(withMock = false) {
+  // Make sure this is checked before `createServer` since it overrides part of
```
Comment (Meital-Kozhidov) [inline be/__tests__/utils.ts:97]: Maybe this comment should come as jsdoc description

File: be/__tests__/utils.ts [hunk 21]
```diff
@@ -271,105 +280,45 @@ export async function cleanupDatabase(db: Database) {
...
-    mode: mode,
-    db: db,
-    mq: mq,
-    cache: cache,
-    allowedOrigins: allowedOrigins,
-    routes: routes,
-    logger: loggerHandler
-  });
+function mockRateLimiter() {
+  vi.spyOn(RateLimiter.prototype, 'getAuthRateLimiter').mockImplementation(
```
Comment (Meital-Kozhidov) [inline be/__tests__/utils.ts:289]: Can you elaborate on the use of `spyOn`? and did you read this https://github.com/vitest-dev/vitest/discussions/4224

File: be/__tests__/utils.ts [hunk 22]
```diff
@@ -12,32 +12,36 @@ import {
...
+  ServerEvents
 } from '../src/event-emitters/index.js';
-import { HttpServer } from '../src/server/index.js';
+import {
+  createCache,
+  createDatabase,
+  createHttpServer,
+  createMessageQueue,
+  createRateLimiter
+} from '../src/init.js';
```
Comment (Meital-Kozhidov) [inline be/__tests__/utils.ts:27]: Yay

File: be/src/db/handler.ts (hunk 5 above)
Comment (Meital-Kozhidov) [reply be/src/db/handler.ts:131]: liked this addition that's all

File: be/src/event-emitters/message-queue.ts (hunk 11 above)
Comment (Meital-Kozhidov) [reply be/src/event-emitters/message-queue.ts:150]: Interesting, sounds to me like this shouldn't be part of the code. Maybe only in dev, but then you have a code bloat for that small thing..

File: be/src/event-emitters/socket-server.ts (hunk 12 above)
Comment (Meital-Kozhidov) [reply be/src/event-emitters/socket-server.ts:95]: image
You're not the nicest today, one might say

File: be/package.json (hunk 16 above)
Comment (AdamAkiva) [reply be/package.json:41]: Faster than the standard stringify, not much else to it

File: be/src/event-emitters/socket-server.ts (hunk 12 above)
Comment (AdamAkiva) [reply be/src/event-emitters/socket-server.ts:95]: I'm having such fucked up and bad days in the last few weeks, I simply can't take it anymore.
Today I were in bed/slept for most of the day, probably will be the same tomorrow (at least I didn't drink)
I'm sorry if you take it personally, I hope you remember that I did (at some point) had better and more pleasant days.
In regards to your question I think it's different again in a later branch, but I'm not sure

File: be/__tests__/utils.ts (hunk 20 above)
Comment (AdamAkiva) [reply be/__tests__/utils.ts:97]: It is much different in later branches, much of the implementation changed

File: be/__tests__/utils.ts (hunk 22 above)
Comment (AdamAkiva) [reply be/__tests__/utils.ts:27]: As usual you are right

File: be/__tests__/integration/auth.test.ts (hunk 18 above)
Comment (Meital-Kozhidov) [reply be/__tests__/integration/auth.test.ts:62]: Just the desire to learn

File: be/__tests__/utils.ts (hunk 21 above)
Comment (AdamAkiva) [reply be/__tests__/utils.ts:289]: I have read it now.
The idea here is that I wanted to swap only specific functions in the implementation.
It can be done using module mock and to be honest I'm not that well versed with the difference.
I don't use a lot of the functionality of mocks (like checking if the function was called and with what params), I usually use it to swap a heavy operation with something more trivial (at least in these cases).
Therefore I'm not sure there is that much of a difference, I will look into it

File: be/src/utils/rate-limiter.ts (hunk 17 above)
Comment (Meital-Kozhidov) [reply be/src/utils/rate-limiter.ts:63]: Maybe I'm reading the link wrong, but it seems to me like the only reason for this function is that return line... where else is the key generation?

File: be/src/event-emitters/message-queue.ts (hunk 11 above)
Comment (AdamAkiva) [reply be/src/event-emitters/message-queue.ts:150]: I'll think about it.
In general as long as it an exception is not thrown it is not that much of an overhead (around and `if`(s) worth) and the code we don't see (internal libraries) add much more.
My thoughts on these matters in general is that if an error occurred we can take a performance hit in addition since the operation failed anyhow

File: be/src/event-emitters/socket-server.ts (hunk 12 above)
Comment (Meital-Kozhidov) [reply be/src/event-emitters/socket-server.ts:95]: I hope you will feel better soon. I'm here if you need anything. Will try not to take it personally

File: be/src/utils/rate-limiter.ts (hunk 17 above)
Comment (AdamAkiva) [reply be/src/utils/rate-limiter.ts:63]: You are not wrong, this is the main reason.
I kept it to remind myself about handing the reverse proxy issue if and when it happens.
I assume that after it is deployed a lot of this will be changed according to what we will discover

File: be/package.json (hunk 16 above)
Comment (Meital-Kozhidov) [reply be/package.json:41]: Interesting to added a dependency for that..

File: be/__tests__/integration/auth.test.ts (hunk 18 above)
Comment (AdamAkiva) [reply be/__tests__/integration/auth.test.ts:62]: I'm sorry if I'm a bit harsh, I'm just so exhausted.
You can also probably attempt to change it in the code and run the tests no?

File: be/package.json (hunk 16 above)
Comment (AdamAkiva) [reply be/package.json:41]: I looked it up beforehand.
This is the [implementation](https://github.com/epoberezkin/fast-json-stable-stringify/blob/master/index.js)
and this is the [npm page](https://www.npmjs.com/package/fast-json-stable-stringify?activeTab=readme)

File: be/src/event-emitters/socket-server.ts (hunk 12 above)
Comment (AdamAkiva) [reply be/src/event-emitters/socket-server.ts:95]: Thanks, you know I'll never use it however and I do apologize for my behavior.
Hopefully it will all be over soon

File: be/__tests__/integration/auth.test.ts (hunk 18 above)
Comment (Meital-Kozhidov) [reply be/__tests__/integration/auth.test.ts:62]: Yes. It does work. I thought you might have something to say about practice, convention, etc.

File: be/src/server/http.ts (hunk 14 above)
Comment (AdamAkiva) [reply be/src/server/http.ts:118]: The reason I'm letting you do it is because I'm not on Ubuntu and don't have VSCode available

Comment (Meital-Kozhidov) [reply be/src/server/http.ts:118]: image
I'm sorry, I might be dumb, but this didn't help

File: be/__tests__/integration/auth.test.ts (hunk 18 above)
Comment (AdamAkiva) [reply be/__tests__/integration/auth.test.ts:62]: I should at some point go over tests best practices but I'm more in the mind of doing something and putting most of the focus on the source code currently.
I need to be better and have a better mindset in regards to this, but I'm just so unmotivated, even from programming that I don't see it happening soon.
I do hope I'll get back to it ASAP but I'm not optimistic about it at all

File: be/src/event-emitters/message-queue.ts (hunk 11 above)
Comment (Meital-Kozhidov) [reply be/src/event-emitters/message-queue.ts:150]: Makes sense

File: be/__tests__/integration/auth.test.ts (hunk 18 above)
Comment (Meital-Kozhidov) [reply be/__tests__/integration/auth.test.ts:62]: :(

File: be/src/server/http.ts (hunk 14 above)
Comment (AdamAkiva) [reply be/src/server/http.ts:118]: More likely that I don't remember the correct location.
Keep this issue open, tomorrow when (at some point) I'll be on the other OS I'll find the part which indicate it and link it to you

Comment (AdamAkiva) [reply be/src/server/http.ts:118]: Yes, it was my mistake the relevant doc is for the `address()` function.
These are the docs:
```js
/**
         * Returns the bound `address`, the address `family` name, and `port` of the server
         * as reported by the operating system if listening on an IP socket
         * (useful to find which port was assigned when getting an OS-assigned address):`{ port: 12346, family: 'IPv4', address: '127.0.0.1' }`.
         *
         * For a server listening on a pipe or Unix domain socket, the name is returned
         * as a string.
         *
         * ```js
         * const server = net.createServer((socket) => {
         *   socket.end('goodbye\n');
         * }).on('error', (err) => {
         *   // Handle errors here.
         *   throw err;
         * });
         *
         * // Grab an arbitrary unused port.
         * server.listen(() => {
         *   console.log('opened server on', server.address());
         * });
         * ```
         *
         * `server.address()` returns `null` before the `'listening'` event has been
         * emitted or after calling `server.close()`.
         * @since v0.1.90
         */
```
Since the different return options are null (if it was called before the listening event), string (if this is a pipe or unix socket), or `AddressInfo` otherwise, we can assert it as `AddressInfo`

**Review by Meital-Kozhidov — APPROVED** [review]
